Implement shape parameter calibration for ClimWIP

lukasbrunner opened the pull request:
**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] Give this pull request a descriptive title that can be used as a one line summary in a changelog
-   [x] Make sure your code is composed of functions of no more than 50 lines and uses meaningful names for variables
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes

Modified recipe/diagnostic

-   [x] Update documentation for the recipe to the `doc/sphinx/source/recipes` folder
-   [x] Update provenance information if needed
-   [x] Assign the author(s) of the affected recipe(s) as reviewer(s)

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #1902
Documentation: https://esmvaltool--1905.org.readthedocs.build/en/1905/recipes/recipe_climwip.html

lukasbrunner commented:
Hey @Peter9192 ! I now have a first running version. If you have time tomorrow at some point we could discuss it. It is quite similar to my pseudo code idea with the addition of the scipy.minimize

Peter9192 commented:
Hey @lukasbrunner I'm on it now. I totally see why you made the utilities file, but it's quite hard to compare the changes now. Do you think you could split up the commits a bit more logically (maybe move everything first without changes, then the actual changes)? That would make it much easier to go commit by commit and evaluate the relevant changes only.

lukasbrunner commented:
Ah sorry yes! I always get ahead of myself and then there are no atomic commits any more... but really the changes to climwip.py are limited to moving functions to utilities.py and utilities.py contains only functions from the original climwip.py. So the only real changes isolated in calibrate_sigmas.py (except for the call in climwip.py and a very small piece cross-checking if sigmas are already set)

Do you still want me to redo this? Might be a good learning process for me anyway so that would be okay :)

Peter9192 commented:
Let's discuss it first

Peter9192 commented on esmvaltool/diag_scripts/weighting/calibrate_sigmas.py:
```diff
@@ -0,0 +1,282 @@
+    # not be provided from this (I think). Use a flag to chose between this and speed?
+    res = minimize(
+        evaluate_future,
+        x0=overall_performance.mean(),
```
Shouldn't x0 be the initial guess for sigma?

Peter9192 commented on esmvaltool/diag_scripts/weighting/calibrate_sigmas.py:
```diff
@@ -0,0 +1,282 @@
+    return res.x
+
+
+def calibrate_sigmas(independence_sigma: Union[float, str, None],
```
I think we need to think a bit more about the 'user interface' in the recipe. Having three potential different options for the input here is a bit too much maybe? Perhaps another input parser could make sense?

Peter9192 commented on esmvaltool/diag_scripts/weighting/calibrate_sigmas.py:
```diff
@@ -0,0 +1,282 @@
+
+    confident = inside_ratio_weighted >= inside_ratio_reference
+    # either the smallest confident sigma or the largerst sigma in total wins
+    if confident:
+        return performance_sigma
+    return 9999 - performance_sigma
+
```
Apologies for cross posting :yum: 

Maybe this is still too simple, but I wanted to offer a thought. If we specify `confidence` as a spectrum around 0.80, where anything below 0.80 is overconfident (and lower is worse), and anything above 0.80 as 'underconfident' (diffident? ; higher is worse), then minimizing something like `abs(confidence - target_confidence)` where `target_confidence=0.80` could maybe work?

lukasbrunner commented:
Hey @Peter9192 !
My next try with what I think is the most basic version of the sigma calibration :) I'm aware that its not quite polished now and the code still contains a couple of open questions. 

- I now use brute instead of minimize. I realize that this might be slightly less elegant but it is more practicable because it allows a plot of what is going on. I think this is quite important to evaluate if the perfect model test did work. In addition it makes the exact definition of the cost function less important. 
- Not sure if my method of extracting the values from the parallel loop inside brute using dict is quite how this should be done... 
- I realize the plot is a bit busy and not easy to interpret but I tried to include all the relevant information and I can add a description to the docs
- I added a new recipe for running the sigma calibration *_test_performance_sigma.yml (the old one did not have enough model for meaningful results) and renamed the old one into *_test_basic.yml. Not sure if this is something that is used in ESMValTool but in the end I think it needs to be clear that both of them are exactly that: toy examples with out much real meaning. Once Ruth is done we can add her recipe as 'real' application. 

**Example figure sigma calibration**

![performance_sigma_calibration](https://user-images.githubusercontent.com/12111171/100903892-b46ad000-34c6-11eb-84a4-4634cc91281a.png)
Figure: Ratio of perfect models inside of the weighted 80% range (thick black line) depending on the performance sigma value. The smallest not-overconfident sigma is marked by the dashed vertical line.  Ratio of weighted/unweighted 80% ranges (min, mean, max; gray line and shading) depending on the performance sigma value. Cost function for finding the ideal sigma value (red, right axis). 

Additional information:
I also show the reference we try to reach (dashed horizontal line) and the unweighted baseline (doted horizontal line). The difference between the two is that for the baseline the perfect model is excluded from the distribution meaning baseline <= reference and baseline = reference for number of models -> infinity. In addition the ratio of perfect models -> unweighted baseline for sigma -> infinity

I also show a measure of "sharpness", i.e., the ratio between weighted and unweighted spread for each perfect model. The smaller the value the stronger the constraint of the weighting, with sharpness -> 1 for sigma -> infinity. 

The cost function needs to basic properties: If no confident sigma can be found it needs to pick the largest possible value (weakest weighting) and for all confident values it needs to pick the smallest possible value.

lukasbrunner commented:
Okay quick summary I've cleaned up the code and adapted the documentation as well as added a plot. It runs through for me using the 'recipe_climwip_test_performance_sigma.yml' and everything seems correct. I'm still happy to talk about some code style choices, in particular I'm not sure about the way I create the plot in 'calibrate_sigmas.py'

@ruthlorenz can you have a look if this runs for you as well? (I use all the models from your case but two which throw an error. And I removed some of the large ensembles to gain a bit of speed - we can maybe think about removing some more models since this is only supposed to be a test case)

@stefsmeets I saw that you assigned yourself as reviewer so if you have time I'd be happy if you can  have a look

(For cross checking: the sigma value I get is 0.64)

stefsmeets commented:
> @stefsmeets I saw that you assigned yourself as reviewer so if you have time I'd be happy if you can have a look

Yep, @Peter9192 is on vacation, so I assigned myself to keep an eye on this one. I will have a look this week 👍

ruthlorenz commented:
I get
`TypeError: cannot compare cftime.DatetimeProlepticGregorian(2096, 1, 16, 12, 0, 0, 0) and cftime.DatetimeGregorian(1982, 1, 16, 12, 0, 0, 0) (different calendars)`
for EC-Earth3, I'll try without that and see.

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,271 @@
+
+    sharpness.attrs['units'] = '1'
+    confidence['sharpness'] = sharpness
+    # ---
```
```suggestion
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,271 @@
+                         coords={'sigma': sigmas},
+                         attrs={'units': '1'})
+    confidence['cost_function'] = costf
+    # ---
```
```suggestion
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -54,14 +54,31 @@ def distance_matrix(values: 'np.ndarray',
 
-    The independence is calculated as a distance matrix between the
-    datasets defined in the `data_array`. Returned is a square matrix
+    Distances are calculated as the area weighted eclidean distance
```
```suggestion
    Distances are calculated as the area weighted euclidean distance
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -54,14 +54,31 @@ def distance_matrix(values: 'np.ndarray',
+    ----------
+    data_array : array_like, shape (N,...)
+        Array of (2 dimensional) model fields.
+    dimn : string
```
dimn = dimension? It is clearer to write it out, imo.

```suggestion
    dimension : string
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -54,14 +54,31 @@ def distance_matrix(values: 'np.ndarray',
+    distances : array_like, shape (N, N)
+        Symmetric matrix of pairwise model distances.
     """
+    assert dimn != 'model_ensemble', 'dimn must not be "model_ensemble"'
```
```suggestion
    assert dimension != 'model_ensemble', 'dimn must not be "model_ensemble"'
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -71,13 +88,13 @@ def calculate_independence(data_array: 'xr.DataArray') -> 'xr.DataArray':
         input_core_dims=[['model_ensemble', 'lat', 'lon'],
                          ['model_ensemble', 'lat', 'lon']],
-        output_core_dims=[['perfect_model_ensemble', 'model_ensemble']],
+        output_core_dims=[[dimn, 'model_ensemble']],
```
```suggestion
        output_core_dims=[[dimension, 'model_ensemble']],
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -71,13 +88,13 @@ def calculate_independence(data_array: 'xr.DataArray') -> 'xr.DataArray':
     diff.attrs['variable_group'] = data_array.name
     diff.attrs["units"] = data_array.units
-    diff['perfect_model_ensemble'] = diff.model_ensemble.values
+    diff[dimn] = diff.model_ensemble.values
```
```suggestion
    diff[dimension] = diff.model_ensemble.values
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -108,9 +120,32 @@ def compute_overall_mean(dataset: 'xr.Dataset',
+    dataset : None or data_array, shape (N,) or (N, N)
+        A vector containing model-observations distances or a matirx containing
+        model-model distances.
+    dimns : string or list of up to two strings
```
It is clearer to write dimensions instead of dimns, imo.

(or perhaps dim / dims if you prefer to stick with xarray nomenclature)

```suggestion
    dimensions: string or list of up to two strings
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -162,19 +196,52 @@ def calculate_weights(
     """
     if performance is not None:
```
To me it is a bit clearer if you set the default values for numerator/denominator/not_nan first, and then change them depending if independence/performance is specified. You can also argue if independence / performance are mutually exclusive parameters whether they deserve to have their own functions to make this distinction clear.

```python
numerator = 1
not_nan = True
denominator = 1

if performance is not None:
        numerator = np.exp(-((performance / performance_sigma)**2))
        # nans in the performance vector indicate models to be excluded
        not_nan = np.isfinite(performance)

if independence is not None:
    # don't consider nan models for independence of other models!
    exp = np.exp(-((independence[:, not_nan] / independence_sigma)**2))
    # Note diagonal = exp(0) = 1, thus this is equal to 1 + sum(i!=j)
    denominator = exp.sum(axis=1)
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -162,19 +196,52 @@ def calculate_weights(
-    # Normalize weights
-    weights /= weights.sum()
+def calculate_weights(
+        performance: Union['xr.DataArray',
+                           None], independence: Union['xr.DataArray', None],
+        performance_sigma: Union[float, None],
+        independence_sigma: Union[float, None]) -> 'xr.DataArray':
```
Nice work on adding the type annotations 👍

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -162,19 +196,52 @@ def calculate_weights(
+        performance_sigma: Union[float, None],
+        independence_sigma: Union[float, None]) -> 'xr.DataArray':
+    """Xarray wrapper for calculate_weights_data."""
+    if performance is None:
+        performance_core_dims = []
+    else:
+        performance_core_dims = ['model_ensemble']
+    if independence is None:
+        independence_core_dims = []
+    else:
+        independence_core_dims = ['model_ensemble', 'model_ensemble_reference']
```
You could also do this, to do it in a more 'pythonic' way:

```suggestion
    performance_core_dims = [] if performance is None else ['model_ensemble']
    independence_core_dims = [] if independence is None else ['model_ensemble', 'model_ensemble_reference']
```

stefsmeets commented on esmvaltool/diag_scripts/weighting/climwip/main.py:
```diff
@@ -275,6 +271,12 @@ def main(cfg):
             independence, independence_contributions)
         visualize_and_save_independence(overall_independence, cfg,
                                         model_ancestors)
+        if independence_sigma is None:
+            errmsg = ' '.join([
+                'independence_sigma must be set if independence_contributions',
+                'is set'
+            ])
+            raise NotImplementedError(errmsg)
```
This will also work 😅

```suggestion
        if independence_sigma is None:
            raise NotImplementedError('`independence_sigma` must be set if '
                                      '`independence_contributions` is set')
```

lukasbrunner replied:
right! i need to get used to this :) learned to do it always with join

stefsmeets reviewed (commented):
Hi @lukasbrunner , nice work! Looks good and I like the way you organize your code in different files, because it makes for a bit better separation. 

My comments are mostly cosmetical / small code improvements. But it would be nice if you could take the time to improve on the docstrings a bit to explain what some of the arguments are doing.

lukasbrunner commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,271 @@
+
+
+def optimize_confidence(target, weights_matrix, performance_sigma):
+    """Optimize the performance_sigma for confidence."""
```
will do

ruthlorenz commented:
runs for me with only one ensemble of EC-EARTH3 and I obtain 0.62.

However, I had a case where the result was 2 (I attached the figure and yml, it happened when I excluded all EC-EARTH3 versions) but I could not find the warning about this being an issue in the main_log (its too big to attach here though) only the calculate_weights_climwip/climwip/log.txt
We might need to think about raising this more prominently.
[performance_sigma_calibration.pdf](https://github.com/ESMValGroup/ESMValTool/files/5681137/performance_sigma_calibration.pdf)

[recipe_climwip_test_performance_sigma.yml.txt](https://github.com/ESMValGroup/ESMValTool/files/5681143/recipe_climwip_test_performance_sigma.yml.txt)

stefsmeets commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

lukasbrunner commented:
> runs for me with only one ensemble of EC-EARTH3 and I obtain 0.62.
> 
> However, I had a case where the result was 2 (I attached the figure and yml, it happened when I excluded all EC-EARTH3 versions) but I could not find the warning about this being an issue in the main_log (its too big to attach here though) only the calculate_weights_climwip/climwip/log.txt
> We might need to think about raising this more prominently.
> [performance_sigma_calibration.pdf](https://github.com/ESMValGroup/ESMValTool/files/5681137/performance_sigma_calibration.pdf)
> 
> [recipe_climwip_test_performance_sigma.yml.txt](https://github.com/ESMValGroup/ESMValTool/files/5681143/recipe_climwip_test_performance_sigma.yml.txt)

That's not good. I actually think it would probably good if this would just raise an error (after plotting it)... In the end the user has to decide what to do with it. In this case it seems clear that there is a "good" solution that is just shy of the 80%. Maybe I can work on the cost function a bit more but I really want to avoid having solution <80% accepted without the user being aware of it.

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,374 @@
+    return percentiles_spread, inside_ratio
+
+
+def optimize_confidence(target: 'xr.DataArray', weights_matrix: 'xr.DataArray',
```
This doesn't optimize, just computes the cost function, right? Why not 

```suggestion
def compute_cost_function(target: 'xr.DataArray', weights_matrix: 'xr.DataArray',
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,374 @@
+
+    Returns
+    -------
+    cost_function_value : float
```
I think it would be super useful if we could write a symbolic representation of the actual cost function here. Something like

```md
This function returns the value of the following cost function:

.. math::

    \begin{cases}
        \frac{x^2-x}{x},& \text{if } x\geq 1\\
        0,              & \text{otherwise}
    \end{cases}

lukasbrunner replied:
I redesigned  the cost function a bit after some discussion with @ruthlorenz . Testing it now.

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,374 @@
+    - It is highly recommended to visually inspect the graphical output of this
+    process to ensure the optimisation worked as intended.
+
+    - Using cases where the largest possible sigma value was pickt is not
```
```suggestion
    - Using cases where the largest possible sigma value was picked is not
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,374 @@
+    - Using cases where the largest possible sigma value was pickt is not
+    recommended. This is normally a sign that
+        * not enough models where used
+        * the diagnostics where not well chose with respect to the target
```
```suggestion
        * the diagnostics where not well chosen with respect to the target
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,374 @@
+        'percentile_spread': percentiles_spread,
+        'inside_ratio': inside_ratio}
+
+    ratio_difference = inside_ratio - inside_ratio_reference
+
+    if ratio_difference < 0:
+        return np.abs(ratio_difference) + SIGMA_RANGE[1] - performance_sigma
+    return performance_sigma - SIGMA_RANGE[1]
```
This block looks still a bit opaque to me. Can we make it more readable, e.g. by something like this:

```suggestion
    overconfident = inside_ratio - inside_ratio_reference < 0
    best_compromise = np.abs(ratio_difference) + SIGMA_RANGE[1] - performance_sigma
    reliable_value = performance_sigma - SIGMA_RANGE[1]
    cost =  best_compromise if overconfident else reliable_value
    return cost
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,374 @@
+                                       overall_independence, performance_sigma,
+                                       independence_sigma)
+
+    cost_function_value = optimize_confidence(target, weights_matrix,
```
```suggestion
    cost_function_value = compute_cost_function(target, weights_matrix,
```

Peter9192 reviewed (commented):
Hey @lukasbrunner nice work, great to see so much progress on this! I had a quick look at the cost function and left some comments there. I hope to have another look at it later, but since Stef has already done a first review, I think you should consider new comment from my side as bonus suggestions rather than critical issues ;-)

lukasbrunner commented:
Hi @Peter9192 !

Sorry it took me a while to get back to this but now I'm on it again. I addressed your comments and updated the cost function a bit. The function now raises an error if no confident sigma can be found (Ruth and I thought that the warming will likely get lost in the output as the tool is quite verbose). To force it the user can just set a value based on the plot (that is still created before failing - see example of my output) if really wanted.  

I had to exclude EC-Earth3 from the recipe as I got calendar errors now. Not sure if this is an update in our archive on in the pre-processor. 

Also now some of the continuous integration checks fail but I can't really figure out what the problem is. Could you have a look at this if you have time?

![performance_sigma_calibration](https://user-images.githubusercontent.com/12111171/106027530-422ff980-60cb-11eb-9497-5b970667f313.png)

Peter9192 commented:
Hey Lukas, that's great! I'll try to have a look soon. Merging master into this branch might help to fix the CI errors.

Peter9192 commented:
> Also now some of the continuous integration checks fail but I can't really figure out what the problem is. Could you have a look at this if you have time?

I'd start with merging origin/master into this branch, that often helps. Furthermore, one of the problems is the link to the iris docs changed. That's been fixed [here](https://github.com/ESMValGroup/ESMValTool/pull/2003) but not merged yet, so you could also try merging that branch into this one, and then see if the docs pass (that'll also serve as a review for #2003).

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py:
```diff
@@ -0,0 +1,367 @@
+
+    WARNING
+    -------
+    It is highly recommended to visually inspect the graphical output of this
```
Maybe some people will not read this. I think it would be good to put the warning in the readthedocs page as well.

lukasbrunner replied:
Good point! Will do.

lukasbrunner commented:
Merging helped :) @Peter9192  can you please check the documentation, I realized that the numbering of the figures is wrong - it somehow starts with 4... (also in the master)

Peter9192 commented:
> 
> 
> Merging helped :) @Peter9192 can you please check the documentation, I realized that the numbering of the figures is wrong - it somehow starts with 4... (also in the master)

That's because numbering starts with the first recipe in the list of available recipes and then just continues.

Peter9192 commented on doc/sphinx/source/recipes/recipe_climwip.rst:
```diff
@@ -15,22 +15,36 @@ Available recipes and diagnostics
+    * ``performance_sigma`` xor ``calibrate_performance_sigma``: If ``performance_contributions`` is given exactly one of the two has to be given. Otherwise they can be skipped or not set.
+
+        * ``performance_sigma``: float setting the shape parameter for the performance weights calculation (determined offline).
+        * ``calibrate_performance_sigma``: dictionary setting the performance sigma calibration. Has to contain at least the key-value pair specifying ``target``: ``variable_group``. Optional parameters for adjusting the calibration are not yet implemented. WARNING: It is highly recommended to visually inspect the graphical output of the calibration to check if everything worked as intended. In c ...
```
```suggestion
        * ``calibrate_performance_sigma``: dictionary setting the performance sigma calibration. Has to contain at least the key-value pair specifying ``target``: ``variable_group``. Optional parameters for adjusting the calibration are not yet implemented. WARNING: It is highly recommended to visually inspect the graphical output of the calibration to check if everything worked as intended. In case the calibration fails, the best performance sigma will still be indicated in the figure (see example :numref:`fig_climwip_5` below) but not automatically picked - the user can decide to use it anyway by setting it in the recipe (not recommenced).
```

this should do it

Peter9192 replied:
and maybe we can trim down the line length a bit...

lukasbrunner replied:
ah sorry didn't see that before. You mean manual line breaks? Will do.

lukasbrunner commented:
Nice that's what I tried to do :)

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

lukasbrunner commented:
Is this a model missing again? Is there a way of finding out if there is more models that are not available? You could comment them out in the recipe yourself directly if you want

Peter9192 commented:
> 
> 
> Is this a model missing again? Is there a way of finding out if there is more models that are not available? You could comment them out in the recipe yourself directly if you want

So it seems, I've tried removing it. I'm afraid that there's currently no easy way to get that info, but it's certainly somewhere on a long list of things that we might wanna do. Let's see if this helps

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
I'll just keep trying ...

@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

The previous missing one was the last dataset, so fingers crossed.. I'm done for the day.

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

valeriupredoi commented:
@Peter9192 when you done abusing the bot, could you pls post a comment about the readiness of this - could it make it through to v2.2 (be ready by the code freeze deadline of Feb 15)? and if so add the release label pls mate. Cheers :beer:

Peter9192 commented:
Yes I think this is as good as it's gonna get but I wanted to confirm that it runs with the bot and hopefully @ruthlorenz has some time to do the scientific check. I'll add the release label 👍

Peter9192 commented:
@esmvalbot please run recipe_climwip_test_performance_sigma.yml

Peter9192 commented:
@ruthlorenz would you have time to do the scientific review this week? If you can try if it runs for you on the ETH resources, then I can stop trying with the bot..

ruthlorenz commented:
yes I can do a scientific review

ruthlorenz commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -108,9 +120,33 @@ def compute_overall_mean(dataset: 'xr.Dataset',
+    Parameters
+    ----------
+    dataset : None or data_array, shape (N,) or (N, N)
+        A vector containing model-observations distances or a matirx containing
```
matrix instead matirx

ruthlorenz commented on esmvaltool/diag_scripts/weighting/climwip/core_functions.py:
```diff
@@ -121,31 +157,33 @@ def combine_ensemble_members(
         Array specifying the model performance. None is mutually exclusive
-        with independence being None.
+        with independence being None. Single values in performance can be
+        nan, then they will be excluded from the indepence calculation as
```
independence instead indepence

ruthlorenz commented on doc/sphinx/source/recipes/recipe_climwip.rst:
```diff
@@ -114,7 +145,7 @@ multiple datasets.
 ----------
 
-* `Brunner et al. 2020 (accepted), Earth Syst. Dynam., <https://doi.org/10.5194/esd-2020-23>`_.
+* `Brunner et al. 2020, Earth Syst. Dynam., <https://doi.org/10.5194/esd-11-995-2020>`_.
```
maybe add Lorenz et al 2018, Knutti et al 2017 to references? they are in references in the scripts, and even if method changed a bit in meantime, the main method is still the same.
and Brunner et al 2020 needs updating with volume and pages.

ruthlorenz commented on esmvaltool/recipes/recipe_climwip_test_performance_sigma.yml:
```diff
@@ -0,0 +1,190 @@
+      standardize: false
+
+diagnostics:
+  calculate_weights_climwip:
```
diagnostics misses description, and provenance tags (themes, realms).
E.g.
```
description: Calculate weights based on performance and independence
themes:
    - phys
realms:
    - atmos
```

lukasbrunner replied:
I pretty much copied that, I have to admit I don't really understand the provenance tags yet

ruthlorenz commented on esmvaltool/recipes/recipe_climwip_test_performance_sigma.yml:
```diff
@@ -0,0 +1,190 @@
+          pr_CLIM: 1
+          psl_CLIM: 1
+
+  weighted_temperature_graph:
```
misses description and provenance tags.
```
description: Plot timeseries of all included models including interquartile range for weighted and unweighted case.
themes:
    - phys
realms:
    - atmos
```

ruthlorenz commented on esmvaltool/recipes/recipe_climwip_test_performance_sigma.yml:
```diff
@@ -0,0 +1,190 @@
+        ancestors: [calculate_weights_climwip/climwip, tas]
+        weights: 'weights.nc'
+
+  weighted_temperature_map:
```
misses description and provenance tags.
```
description: Plot map of weighted multi-model mean and difference between weighted and unweighted multi-model mean.
themes:
    - phys
realms:
    - atmos
```

ruthlorenz reviewed (changes requested):
I added a few little things that need changing. Otherwise looks good to me and runs fine.
However, I am not completely independent since I run on the same servers as @lukasbrunner.

lukasbrunner commented:
Thanks @ruthlorenz for the careful check!! 
@Peter9192 not sure why CI fails now I only touched comments since my earlier commits. (except maybe the added information to the recipe?) Or the merge of the master?

Peter9192 commented:
Seems unrelated to the PR, something with the test_cmorize_obs. Guessing this might be due to some of the refactoring of the logger we did recently in the core (@stefsmeets ?).

stefsmeets commented:
> Seems unrelated to the PR, something with the test_cmorize_obs. Guessing this might be due to some of the refactoring of the logger we did recently in the core (@stefsmeets ?).

I don't know what happened with this, should have been fixed by https://github.com/ESMValGroup/ESMValTool/pull/2020

valeriupredoi commented:
#2020 fixed it for the current development `master` branch of ESMValCore, but broke it for the released version which is used for the CI tests stirring the wrath of @bouweandela while at it :grin:

Peter9192 commented:
One more minor technical remark @lukasbrunner, can you run yamllint on the changed recipes? I think you might want to trim down the line length on some of the descriptions you've added.

Peter9192 commented:
Thanks! 
Since the CI issue has nothing to do with this PR I'm happy to see this get merged

bouweandela commented:
Could you please merge the master branch into this pull request, so all the tests pass?

valeriupredoi commented:
I take it that @ruthlorenz ok-ed this scientologically and @Peter9192 technically right? :grin: I am going to change labels and pending test passing I am going to merge, please object if otherwise :+1:

ruthlorenz reviewed (approved):
All good now
